skypy/halo/mass.py

```diff
@@ -1,38 +1,264 @@
-r'''Halo mass module.
-
-This module provides methods to sample the  masses of dark matter halos.
+r'''Halo mass sampler.
+This code samples halos from their mass function.
 
 Models
 ======
-
 .. autosummary::
    :nosignatures:
    :toctree: ../api/
 
    press_schechter
-
+   halo_mass_function
+   halo_mass_sampler
+   sheth_tormen_collapse_function
+   press_schechter_collapse_function
 '''
 
 import numpy as np
+from scipy import integrate
+from functools import partial
+from astropy import units as u
 
 from skypy.utils.random import schechter
 
+__all__ = [
+    'press_schechter',
+    'halo_mass_function',
+    'halo_mass_sampler',
+    'sheth_tormen_collapse_function',
+    'press_schechter_collapse_function',
+ ]
+
+
+def halo_mass_function(M, sigma, collapse_function, cosmology):
+    r'''Halo mass function.
+    This function computes the halo mass function, defined
+    in equation 7.46 in [1]_.
+
+    Parameters
+    -----------
+    M : (nm,)
+        Array for the halo mass, in units of solar masses.
+    sigma: (ns,) array_like
+        Array of the mass variance at different scales and at a given redshift.
+    collapse_function: (nm,) array_like
+        Collapse function to choose from a variety of models:
+        `sheth_tormen_collapse_function`, `press_schechter_collapse_function`.
+    cosmology : astropy.cosmology.Cosmology
+        Cosmology object providing methods for the evolution history of
+        omega_matter and omega_lambda with redshift.
+
+    Returns
+    --------
+    mass_function: (nm,) array_like
+        Halo mass function for a given mass array, cosmology and redshift, in
+        units of :math:`Mpc^{-3} M_{Sun}^{-1}`.
+
+    Examples
+    ---------
+    >>> import numpy as np
+    >>> from skypy.halo import mass
+    >>> from skypy.power_spectrum import _eisenstein_hu as eh
+    >>> from skypy.power_spectrum import growth_function
+
+    This example will compute the halo mass function for spherical collapse
+    and a Planck15 cosmology. The power spectrum is given by the Eisenstein
+    and Hu fitting formula:
+
+    >>> from astropy.cosmology import Planck15
+    >>> cosmology = Planck15
+    >>> redshift = 0.0
+    >>> k = np.logspace(-3, 1, num=5, base=10.0)
+    >>> A_s, n_s = 2.1982e-09, 0.969453
+    >>> Pk = eh.eisenstein_hu(k, A_s, n_s, cosmology, kwmap=0.02, wiggle=True)
+    >>> D0 = growth_function(0, cosmology)
+    >>> Dz = growth_function(redshift, cosmology) / D0
+
+    And the Press-Schechter mass function is computed
+
+    >>> m = 10**np.arange(9.0, 12.0, 2)
+    >>> sigma = np.sqrt(_sigma_squared(m, k, Pk, 0, cosmology))
```

Comment:
```suggestion
    >>> sigma = np.sqrt(_sigma_squared(m, k, Pk, Dz, cosmology))
```
I think this is why one of your tests is failing. Setting `Dz = 0` leads to `sigma = 0` and then you divide by zero in `_dlns_dlnM` to get a `NaN`